text-sregex.c

```diff
@@ -0,0 +1,146 @@
+#include <stdlib.h>
+#include <string.h>
+#include <sregex.h>
+
+#include "text-regex.h"
+
+struct Regex {
+    sre_pool_t *pool; // pool for parsing and compiling
+    sre_pool_t *mpool; // pool for matching
+    sre_regex_t *regex;
+    sre_program_t *prog;
+    sre_uint_t ncaptures;
+    sre_int_t* captures;
+};
+
+static sre_uint_t regex_capture_size(Regex *r) {
+    return 2 * sizeof(sre_int_t) * (r->ncaptures + 1);
+}
+
+static int translate_cflags(int cflags) {
+    return 0;
+}
+
+static sre_vm_pike_ctx_t *make_context(Regex *r) {
+    return sre_vm_pike_create_ctx(r->mpool, r->prog, r->captures, regex_capture_size(r));
+}
+
+static void destroy_context(sre_vm_pike_ctx_t *ctx, Regex *r) {
+    sre_reset_pool(r->mpool);
+}
+
+static void fill_match(Regex *r, size_t nmatch, RegexMatch pmatch[], size_t offset) {
+    sre_int_t *cap = r->captures;
+    for (size_t i = 0; i < nmatch; i++) {
+        pmatch[i].start = *cap == -1 ? EPOS : *cap + offset;
+        ++cap;
+        pmatch[i].end = *cap == -1 ? EPOS : *cap + offset;
+        ++cap;
+    }
+}
+
+Regex *text_regex_new(void) {
+    Regex* result = calloc(1, sizeof(Regex));
+    return result;
+}
+
+int text_regex_compile(Regex *r, const char *string, int cflags) {
+    sre_int_t err_offset;
+    r->pool = sre_create_pool(4096);
+
+    if (!r->pool) return 1;
+    r->regex = sre_regex_parse(r->pool, (sre_char *) string, &r->ncaptures, translate_cflags(cflags), &err_offset);
+    if (!r->regex) goto err1;
+    r->prog = sre_regex_compile(r->pool, r->regex);
+    if (!r->prog) goto err1;
+    r->captures = malloc(regex_capture_size(r));
+    if (!r->captures) goto err1;
+    r->mpool = sre_create_pool(4096);
+    if (!r->mpool) goto err1;
+    return 0;
+
+err1:
+    sre_destroy_pool(r->pool);
+    return 1;
+}
+
+void text_regex_free(Regex *r) {
+    if (r) {
```

Comment:
The general idiom in vis is to start these `*_free` functions with `if (!r) return;` to avoid one level of indentation. It isn't really that important here because the function is extremely short.